### DISCUSSION TITLE
Setting up email inside Docker Container

### DISCUSSION THREAD

**Discussion by okynnor** [alexjustesen/speedtest-tracker#282] [Q&A]
I'm sure this going to be obvious. 
I'm running it in a Docker container.  

Per the docs, it says how to setup the email server for notification in the .env file inside the config.  I can't seem to locate it.  Can you add in the docs and let me know? 

https://docs.speedtest-tracker.dev/settings/notifications/mail

`.env file in your mounted data directory.`

I'm so glad you guys are making a new speedtest.  The old one just died and was unreliable at the least. Keep up the great work! It's coming a long greatly.

**Comment by alexjustesen** [comment]
The `.env` file will be in the mounted `/config` directory you setup during initial install. The app creates this file on the first boot with a bunch of default settings.

https://docs.speedtest-tracker.dev/getting-started/installation#install-with-docker

**Reply by alexjustesen** [reply]
Also thanks for the support! Just getting started and have a lot of ideas on how to make it a better app in the future.

**Reply by alexdelprete** [reply]
> The app creates this file on the first boot with a bunch of default settings.

In the doc I would also specify that all environment variables in the .env file can also be configured directly in the docker-compose file (if used). The .env file is simply a central file for the environment variables that Docker reads when starting up the container.

Reading the docs it seems that the .env file is mandatory, but that is not the case.

![image](https://user-images.githubusercontent.com/7027842/210238539-d99a5176-bb28-4b39-a42c-eb34944db285.png)

All those variables can be configured in the docker-compose file.

**Comment by okynnor** [comment]
I honestly didn't see an .env file in the /config folder so I opted for adding it to the Docker compose. Adding the mail server setting note to the Doc would be very helpful.  For me, I added them to my docker compose (or as they are known in Portainer as "Stacks"). 

I got an internal server error 500.  The error is not informative to say the least. 

I'm pretty sure everything else is correct except for the format for the MAIL_Encryption for Gmail.  Is it tls, tls/ssl or just ssl?  I've tried them all to no avail.

**Reply by alexdelprete** [reply]
This works for Gmail SMTP (see #176):

```
MAIL_MAILER=smtp
MAIL_HOST=smtp.gmail.com
MAIL_PORT=587
MAIL_USERNAME="*****@gmail.com"
MAIL_PASSWORD="******************"
MAIL_ENCRYPTION=tls
MAIL_FROM_ADDRESS="*****@gmail.com"
MAIL_FROM_NAME="${APP_NAME}"
```

> I got an internal server error 500. The error is not informative to say the least.

I don't understand where you get error 500, when pressing the test email button? Be specific please. If it's the test button, please make sure the recipient list is not empty.

**Reply by okynnor** [reply]
I tried your suggested settings for mail.  I changed the port from 465 to yours 587.  I added MAIL_MAILER=smtp and change MAIL_FROM_NAME to match yours, all of which weren't in the doc.  Unfortunately, they didn't work either. 

Here's a screenshot of my GUI under Notifications to show you. So, I would hit "test mail channel" and I would see the 500 - Server Error. 

![Notifications_-_Speedtest_Tracker](https://user-images.githubusercontent.com/40222847/210392361-58d962e1-6f80-41d3-8135-5678948611ec.png)

**Reply by alexdelprete** [reply]
Can you please show me the file with all the settings please? I think @alexjustesen fixed this 500 error, it was due to the empty recipient list.

**Reply by okynnor** [reply]
If you would refer to my screenshot earlier, I pixelated out my email address...I mean that my recipient email address field is populated before I pressed on the button "Test mail channel".  

As per your request: 

`version: '3.3'
services:
    speedtest-tracker:
        container_name: speedtest-tracker
        ports:
            - '8080:80'
            - '8443:443'
        environment:
            - PUID=1000
            - PGID=1000
            - DB_CONNECTION=mysql
            - DB_HOST=db
            - DB_PORT=3306
            - DB_DATABASE=speedtest_tracker
            - DB_USERNAME=speedy
            - DB_PASSWORD=password
            - MAIL_MAILER=smtp
            - MAIL_HOST=smtp.gmail.com
            - MAIL PORT=465
            - MAIL USERNAME=my.username
            - MAIL PASSWORD=my.googlepassword.here
            - MAIL ENCRYPTION=tls
            - MAIL_FROM_NAME="${APP_NAME}"
        volumes:
            - '/var/lib/docker/volumes/Configs:/config'
        image: 'ghcr.io/alexjustesen/speedtest-tracker:latest'
        restart: unless-stopped
        depends_on:
            - db
    db:
        image: mariadb:10
        restart: always
        environment:
            - MARIADB_DATABASE=speedtest_tracker
            - MARIADB_USER=speedy
            - MARIADB_PASSWORD=password
            - MARIADB_RANDOM_ROOT_PASSWORD=true
        volumes:
            - speedtest-db:/var/lib/mysql
volumes:
  speedtest-db:`

**Reply by alexjustesen** [reply]
@okynnor make sure you click the "add to recipients" button first then "save" before click "test".

Note to self, this isn't obvious and probably needs to be smoothed out.

**Reply by alexjustesen** [reply]
You can also change `APP_DEBUG` in `.env` to `true` to get more verbose errors.

**Reply by okynnor** [reply]
Here's the error with the `

![Screen Shot 2023-01-03 at 13 08 58](https://user-images.githubusercontent.com/40222847/210442150-08f6df6c-c007-4b0d-b1ca-aa30c6d4dcc1.png)` set to true

**Reply by alexjustesen** [reply]
Thanks, I'll take another look at this and see if I can enhance the process too

**Reply by alexjustesen** [reply]
@okynnor heads up the next release contains a fix which should resolve your issue https://github.com/alexjustesen/speedtest-tracker/pull/289. The issue was because you had an empty recipient.

**Comment by okynnor** [comment]
This is great news!  

@alexjustesen As you saw in my screenshot, I had my email address in the GUI. Was it that the variable wasn't being passed to the php code?  Now that's fixed right?

**Reply by alexjustesen** [reply]
It was more so because you had an empty value

**Comment by okynnor** [comment]
I don’t understand how. As long as it’s fixed.

**Comment by okynnor** [comment]
I just downloaded a new version.  Has this notification email issue been fixed?  This is the error I'm still getting as of this reply.  Note: I updated to the latest version today. 

![Screenshot 2023-01-07 at 1 39 49 PM](https://user-images.githubusercontent.com/40222847/211171270-bb21f0ab-678c-44f2-9f50-8b63fa035951.png)
![Screenshot 2023-01-07 at 1 42 30 PM](https://user-images.githubusercontent.com/40222847/211171342-2d955581-4de0-403d-a782-5cc7876805fd.png)
![Screenshot 2023-01-07 at 1 41 53 PM](https://user-images.githubusercontent.com/40222847/211171345-d9b51a99-95f2-4e19-89e6-af9be83cb9c0.png)

**Reply by alexjustesen** [reply]
Remove that 2nd empty value, you should only have boxes where you actually have an email address in the field.

**Comment by okynnor** [comment]
Geez. that was a simple solution to the error.  May I suggest to add code to make it disappear if it's empty and/or not save an empty field? 

NEW: The app shows the message "Test Mail Notification Sent." after pressing "Test Mail Channel" button.  Sadly, no email ever arrives to my email address which is set to my_email_address@yahoo.com.  Not even in the Junk Folder.  Waited for 30 mins. Nothing.  

Since I use Gmail's smtp servers for this, I check Gmail's sent folder.  There was nothing there.  

What can I do to help troubleshoot if gmail's smtp server is even connected? I know it works because I use the same credentials in my other apps.

**Reply by alexjustesen** [reply]
So `v0.10.0` which is building now contains better validation so there won't be an empty email field allowed. Should be available shortly https://github.com/alexjustesen/speedtest-tracker/releases/tag/v0.10.0

**Reply by alexjustesen** [reply]
Also on the email, I "think" I remember some jank happening when dealing with Gmail SMTP and I haven't used it in a while because of that. I switched to Mailgun because you get like 1000 messages / month free.

**Reply by okynnor** [reply]
changing the ports from 465 to 587 doesn't work too.  Must be Gmail support needs to be made later. 

In the meantime, would like to me try and then report whether Mailgun works?
